src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js

```diff
@@ -0,0 +1,198 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview main module for the safemail app.
+ * @author yzhu@yahoo-inc.com (Yan Zhu)
+ */
+
+goog.provide('e2e.openpgp.pgpmime.MimeNode');
+
+goog.require('e2e.openpgp.pgpmime.Constants');
+goog.require('e2e.openpgp.pgpmime.Text');
+goog.require('e2e.openpgp.pgpmime.Utils');
+
+goog.require('goog.array');
+goog.require('goog.asserts');
+goog.require('goog.crypt.base64');
+goog.require('goog.object');
+goog.require('goog.string');
+
+goog.scope(function() {
+var pgpmime = e2e.openpgp.pgpmime;
+var constants = pgpmime.Constants;
+
+
+
+/**
+ * Constructor for a MIME tree node.
+ * @param {{
+ *   multipart: boolean,
+ *   optionalHeaders: (Array<!e2e.openpgp.pgpmime.types.Header>|undefined)
+ * }} options Options to initialize for the node.
+ * @param {e2e.openpgp.pgpmime.MimeNode=} opt_parent The parent node.
+ * @constructor
+ */
+e2e.openpgp.pgpmime.MimeNode = function(options, opt_parent) {
+  this.parent = opt_parent || this;
+  this.multipart_ = options.multipart;
+
+  this.children_ = [];
+  this.header_ = /** @type {e2e.openpgp.pgpmime.types.Header} */ ({});
+  this.content_ = null;
+  this.boundary_ = '';
+
+  // set headers
+  if (goog.isDefAndNotNull(options.optionalHeaders)) {
+    goog.array.forEach(options.optionalHeaders, goog.bind(function(header) {
+      if (goog.isDefAndNotNull(header.name) &&
+          goog.isDefAndNotNull(header.value)) {
+        if (goog.isDefAndNotNull(header.params)) {
+          this.setHeader_(header.name, header.value, header.params);
+        } else {
+          this.setHeader_(header.name, header.value);
+        }
+      }
+    }, this));
+  }
+
+  this.setBoundary_();
+};
+
+
+/**
+ * Sets the MIME message boundary for a node.
+ * @private
+ */
+e2e.openpgp.pgpmime.MimeNode.prototype.setBoundary_ = function() {
+  // TODO: Strictly ensure that the boundary value doesn't coincide with
+  //   any string in the email content and headers.
+  this.boundary_ = '---' + goog.string.getRandomString() +
+      Math.floor(Date.now() / 1000).toString();
+};
+
+
+/**
+ * Adds a child to a MIME node.
+ * @param {{
+ *   multipart: boolean,
+ *   optionalHeaders: (Array<!e2e.openpgp.pgpmime.types.Header>|undefined)
+ * }} options Options to initialize for the node.
+ * @return {e2e.openpgp.pgpmime.MimeNode}
+ */
+e2e.openpgp.pgpmime.MimeNode.prototype.addChild = function(options) {
+  var node = new e2e.openpgp.pgpmime.MimeNode(options, this);
+  this.children_.push(node);
+  return node;
+};
+
+
+/**
+ * Sets a MIME header.
+ * @param {string} key Name of the header.
+ * @param {string} value Value of the header, possibly including parameters.
+ * @param {Object.<string, string>=} opt_params Optional dict of additional
+ *   parameters
```

Comment:
Indent this line 4 spaces, not counting the first space that comes from JSDoc formatting  (i.e. you're missing two spaces).